### PR TITLE
[SPARK-23021][SQL] AnalysisBarrier should override innerChildren to print correct explain output

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
`AnalysisBarrier` in the current master cuts off explain results for parsed logical plans;
```
scala> Seq((1, 1)).toDF("a", "b").groupBy("a").count().sample(0.1).explain(true)
== Parsed Logical Plan ==
Sample 0.0, 0.1, false, -7661439431999668039
+- AnalysisBarrier Aggregate [a#5], [a#5, count(1) AS count#14L]
```
To fix this, `AnalysisBarrier` needs to override `innerChildren` and this pr changed the output to;
```
== Parsed Logical Plan ==
Sample 0.0, 0.1, false, -5086223488015741426
+- AnalysisBarrier
      +- Aggregate [a#5], [a#5, count(1) AS count#14L]
         +- Project [_1#2 AS a#5, _2#3 AS b#6]
            +- LocalRelation [_1#2, _2#3]
```

## How was this patch tested?
Added tests in `DataFrameSuite`.

### CODE REVIEW

**Comment by rednaxelafx** [comment]
Thanks for fixing it!

One thing I'm curious what others think is, whether or not the `AnalysisBarrier` nodes themselves should show up in the explain output at all, i.e. should it behave more like `WholeStageCodegenExec`/`InputAdapter` in physical plans that just hides themselves from the explain output, or just show up as a normal node.

I'm okay either way.

**Comment by viirya** [comment]
Because AnalysisBarrier will be removed in analyzed plan, maybe leaving it in parsed plan is ok. This fix is more simple. But if any one thinks it is necessary to remove it from parsed plan too, I'm also ok.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by maropu** [comment]
@rednaxelafx yea, I'm also ok to print `AnalysisBarrier` cuz this info is some informative.

**Comment by maropu** [comment]
Could you check? @viirya @cloud-fan (btw, I'm bit surprise @viirya moved to fukuoka?)

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -2228,4 +2229,26 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     checkAnswer(df, Row(0, 10) :: Nil)
     assert(df.queryExecution.executedPlan.isInstanceOf[WholeStageCodegenExec])
   }
+
+  test("SPARK-23021 AnalysisBarrier should not cut off explain output for parsed logical plans") {
+    val df = Seq((1, 1)).toDF("a", "b").groupBy("a").count().limit(1)
+    val explain = ExplainCommand(df.queryExecution.logical, extended = true)
+    val explainString =
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: In `StreamSuite`, there is a trick I think it's better to obtain explain command's output:

```scala
val outputStream = new java.io.ByteArrayOutputStream()
Console.withOut(outputStream) {
  joined.explain(true)
}
assert(outputStream.toString.contains("StreamingRelation"))
```

Do you want to try it?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: ok, I'll fix that way

**Comment by viirya** [comment]
@maropu Yeah, I'm in Japan now.

**Comment by kiszk** [comment]
Welcome to Japan

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -2228,4 +2228,21 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     checkAnswer(df, Row(0, 10) :: Nil)
     assert(df.queryExecution.executedPlan.isInstanceOf[WholeStageCodegenExec])
   }
+
+  test("SPARK-23021 AnalysisBarrier should not cut off explain output for parsed logical plans") {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Move it to `HiveExplainSuite`,

**Comment by gatorsmile** [comment]
LGTM except a minor comment.

**Review by viirya — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

**Pull request closed without merging**